agent_build/docker_images/image_builder.py

```diff
@@ -0,0 +1,195 @@
+                continue
+
+            # According to the Dockerfile format, the ADD operator support multiple source paths
+            # and the last is the destination path.
+
+            # Split the ADD or COPY line.
+            add_or_copy_line_parts = line.strip().split(" ")
+
+            # Ignore the multi-stage build COPY operators.
+            if add_or_copy_line_parts[0] == "COPY" and add_or_copy_line_parts[1].startswith("--from"):
+                continue
+
+            # The first part is the operator itself, and the last the destination path.
+            # Everything between - the source files.
+
+            for source_path in add_or_copy_line_parts[1:len(add_or_copy_line_parts) - 1]:
+                # Since we know the path for the image's build context, we can resolve the absolute paths.
+
+                used_paths.append(
+                    pl.Path(self._source_dir, source_path).absolute()
+                )
+
+        # Take into account the dockerfile itself.
+        used_paths.append(self._dockerfile_path)
+        # and this script
+        used_paths.append(pl.Path(__file__))
+
+        # All paths that are used in the image have been determined. But there may be directories, so we also have to
+        # take into account files inside them. The easiest way is to just walk through those directories and get all
+        # files.
+        used_files = []
+
+        def get_dir_files(dir_path: pl.Path):
+            if dir_path.name == "__pycache__":
+                return []
+
+            result = []
+            for child_path in dir_path.iterdir():
+                if child_path.is_dir():
+                    result.extend(get_dir_files(child_path))
+                else:
+                    result.append(child_path)
+
+            return result
+
+        for used_path in used_paths:
+            # skip if not directory.
+            if not used_path.is_dir():
+                used_files.append(used_path)
+                continue
+
+            used_files.extend(get_dir_files(used_path))
+
+        return used_files
+
+    def _build(self, image_name: str):
+        """
+        Build the docker image.
+        """
+        # Before the build, check if there is already an image with the same name. The name contains a checksum of all
+        # content of the image, so the image name has to be the same if content is also the same..
+        # If so, skip the build.
+        output = subprocess.check_output(
+            f"docker images -q {image_name}", shell=True
+        ).decode().strip()
+
+        if output:
+            # The image already exists, skip the build.
+            return
+
+        # Add build args.
+        if self._build_args:
+            build_args = ""
+            for arg_name, arg in self._build_args.items():
+                build_args = f'{build_args} --build-arg {arg_name}="{arg}"'
+        else:
+            build_args = ""
+
+        subprocess.check_call(
```

Comment:
I know it's build only script, but to be on the safe side we should probably use array of args so they are escaped properly or escape them manually.

Reply:
I removed using shell everywhere where there is dynamic input.